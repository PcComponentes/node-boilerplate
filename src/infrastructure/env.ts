export enum EnvMode {
  DEV = 'DEV',
  PROD = 'PROD',
  TEST = 'TEST',
}

export interface EnvKey {
  PORT: string;
  MODE: string;
}

export class Env {
  private env: EnvKey;

  constructor() {
    this.env = {
      PORT: '',
      MODE: EnvMode.DEV,
    };

    const envKeys = Object.keys(this.env) as Array<keyof EnvKey>;

    for (const key of envKeys) {
      const value = process.env[key];

      if (!value) {
        throw new Error(`Missing env key ${key}`);
      } else {
        this.env[key] = value;
      }
    }

    if (Object.keys(EnvMode).indexOf(this.get('MODE')) === -1) {
      throw new Error(`Invalid env MODE value.`);
    }
  }

  public get<T extends keyof EnvKey>(key: T): EnvKey[T] {
    return this.env[key];
  }

  public set<T extends keyof EnvKey>(key: T, value: EnvKey[T]) {
    this.env[key] = value;
  }
}
