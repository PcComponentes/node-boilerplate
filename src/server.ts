import express, { Application } from "express";
import { router } from "./entrypoint/routes";
import { Env } from "./infrastructure/env";

export class Server {
  private app: Application;

  constructor(private env: Env) {
    this.app = express();
    this.registerRoutes();
  }

  private registerRoutes() {
    this.app.use(router);
  }

  public getApp() {
    return this.app;
  }

  public listen() {
    const port = this.env.get("PORT");

    this.app.listen(port, () => {
      return console.log(`Server is listening on ${port}`);
    });
  }
}
