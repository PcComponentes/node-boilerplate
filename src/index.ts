import { Env } from './infrastructure/env';
import { Server } from './server';

const env = new Env();
const server = new Server(env);
server.listen();
