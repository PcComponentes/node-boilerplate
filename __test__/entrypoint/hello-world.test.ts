import { Application } from 'express';
import supertest from 'supertest';
import { Env } from '../../src/infrastructure/env';
import { Server } from '../../src/server';

describe('Hello World API', () => {
  let server: Server;
  let app: Application;

  beforeAll(async (done) => {
    const env = new Env();
    server = new Server(env);
    app = server.getApp();

    done();
  });

  it('should return 200 when getting hello-world', async (done) => {
    const res = await supertest(app).get('/hello-world').send();

    expect(res.status).toBe(200);
    expect(res.text).toMatchSnapshot();

    done();
  });
});
