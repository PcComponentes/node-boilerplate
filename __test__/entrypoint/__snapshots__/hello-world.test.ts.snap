// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Hello World API should return 200 when getting hello-world 1`] = `"{\\"text\\":\\"hello world\\"}"`;
